components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -12,302 +12,603 @@
+/// # };
+/// # use exonum_crypto::hash;
+/// let fork = { let db = TemporaryDB::new(); db.fork() };
+/// let mut list = ProofListIndex::new("index", &fork);
+/// list.extend(vec![100_u32, 200_u32, 300_u32]);
 ///
-/// Proof of absence for an element with the specified index consists of
-/// `merkle_root` of `ProofListIndex` and `length` of the list.
+/// // Get the proof from the index
+/// let proof = list.get_range_proof(1..);
 ///
-/// Element with `index` is absent in the list with provided `length`
-/// and `merkle_root` when two conditions are met:
-/// ```text
-/// 1. list_hash == sha256( HashTag::List || length || merkle_root )
-/// 2. index > length
+/// // Check the proof consistency
+/// let elements = proof.validate(list.object_hash(), list.len()).unwrap();
+/// assert_eq!(*elements, [(1, 200_u32), (2, 300_u32)]);
 /// ```
 ///
-/// In case of a range proof this rule applies to the whole range.
-#[derive(Debug, PartialEq, Clone, Eq)]
-pub struct ProofOfAbsence {
-    length: u64,
-    merkle_root: Hash,
+/// # JSON serialization
+///
+/// `ListProof` is serialized to JSON as an object with 2 array fields:
+///
+/// - `hashes` is an array of `{ height: number, index: number, hash: Hash }` objects.
+/// - `values` is an array with list elements and their indexes, that is,
+///   tuples `[number, V]`.
+///
+/// ```
+/// # use serde_json::{self, json};
+/// # use exonum_merkledb::{Database, TemporaryDB, BinaryValue, HashTag, ListProof, ProofListIndex};
+/// # fn main() {
+/// let fork = { let db = TemporaryDB::new(); db.fork() };
+/// let mut list = ProofListIndex::new("index", &fork);
+/// list.extend(vec![1_u32, 2, 3]);
+/// let h1 = HashTag::hash_leaf(&1_u32.to_bytes());
+/// let h3 = HashTag::hash_leaf(&3_u32.to_bytes());
+/// let h33 = HashTag::hash_single_node(&h3);
+///
+/// let proof = list.get_proof(1);
+/// assert_eq!(
+///     serde_json::to_value(&proof).unwrap(),
+///     json!({
+///         "hashes": [
+///             { "index": 0, "height": 1, "hash": h1 },
+///             { "index": 1, "height": 2, "hash": h33 },
+///         ],
+///         "values": [ [1, 2] ],
+///     })
+/// );
+/// # }
+/// ```
+///
+/// [`get_proof()`]: struct.ProofListIndex.html#method.get_proof
+/// [`get_range_proof()`]: struct.ProofListIndex.html#method.get_range_proof
+/// [`validate()`]: #method.validate
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct ListProof<V> {
+    hashes: Vec<HashedEntry>,
+    values: Vec<(u64, V)>,
 }
 
-impl ProofOfAbsence {
-    /// New `ProofOfAbsence` for specified list `length` and `merkle_root`.
-    pub fn new(length: u64, merkle_root: Hash) -> Self {
-        Self {
-            length,
-            merkle_root,
-        }
+fn merge(
+    first: impl Iterator<Item = HashedEntry>,
+    second: impl Iterator<Item = HashedEntry>,
+) -> impl Iterator<Item = Result<HashedEntry, ()>> {
+    struct Merge<T, U> {
```

Comment:
I've used [naming conventions for iterators](https://doc.rust-lang.org/1.0.0/style/style/naming/iterators.html) - since the function is named `merge`, the iterator is named `Merge` as well.

Reply:
Hm.. Ok. However, I don't really like it in that particular case.